core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java

```diff
@@ -23,86 +23,165 @@
+                "This tokenize synonyms with whatever tokenizer and token filters appear before it in the chain. " +
+                "If you need ignore case with this filter, you should set lowercase filter before this");
+        }
+
+        this.expand =
+            settings.getAsBooleanLenientForPreEs6Indices(indexSettings.getIndexVersionCreated(), "expand", true, deprecationLogger);
+
+        // for backward compatibility
+        if (indexSettings.getIndexVersionCreated().before(Version.V_6_0_0_alpha3)) {
+            deprecationLogger.deprecated(
+                "This filter tokenize synonyms with whatever tokenizer and token filters appear before it in the chain.");
+            String tokenizerName = settings.get("tokenizer", "whitespace");
+            AnalysisModule.AnalysisProvider<TokenizerFactory> tokenizerFactoryFactory =
+                analysisRegistry.getTokenizerProvider(tokenizerName, indexSettings);
+            if (tokenizerFactoryFactory == null) {
+                throw new IllegalArgumentException("failed to find tokenizer [" + tokenizerName + "] for synonym token filter");
+            }
+            final TokenizerFactory tokenizerFactory = tokenizerFactoryFactory.get(indexSettings, env, tokenizerName,
+                AnalysisRegistry.getSettingsFromIndexSettings(indexSettings,
+                    AnalysisRegistry.INDEX_ANALYSIS_TOKENIZER + "." + tokenizerName));
+            this.tokenizerFactory = tokenizerFactory;
+        } else {
+            this.tokenizerFactory = null;
+        }
+
+        this.format = settings.get("format", "");
+    }
+
+    @Override
+    public TokenStream create(TokenStream tokenStream) {
+        throw new IllegalStateException("Call createPerAnalyzerSynonymFactory to specialize this factory for an analysis chain first");
+    }
+
+    protected String getRulesFromSettings(Environment env) {
+        String rules;
         if (settings.getAsArray("synonyms", null) != null) {
-            List<String> rules = Analysis.getWordList(env, settings, "synonyms");
+            List<String> rulesList = Analysis.getWordList(env, settings, "synonyms");
             StringBuilder sb = new StringBuilder();
-            for (String line : rules) {
+            for (String line : rulesList) {
                 sb.append(line).append(System.lineSeparator());
             }
-            rulesReader = new FastStringReader(sb.toString());
+            rules = sb.toString();
         } else if (settings.get("synonyms_path") != null) {
-            rulesReader = Analysis.getReaderFromFile(env, settings, "synonyms_path");
+            Reader rulesReader = Analysis.getReaderFromFile(env, settings, "synonyms_path");
+            StringBuilder sb = new StringBuilder();
+            String tmp;
+            try (BufferedReader br = new BufferedReader(rulesReader)){
+                while ((tmp = br.readLine()) != null) {
+                    sb.append(tmp).append(System.getProperty("line.separator"));
```

Comment:
Please do not load the content of the file in memory. I think it should be easy to do by changing the return type of this method to a `java.io.Reader`?